### PR TITLE
Restore deleted tab context menu

### PR DESCRIPTION
**Pull request by bsclifton**
Redo of https://github.com/brave/brave-core/pull/3836
(review feedback came in after I had merged it)

Fix brave/brave-browser#6513

This reverts commit 750c9a8c4e5cb70fcf01a58b9acc27b2fbfc79d6, reversing
changes made to 077fc771d2901d8f55db267c7bbaa151e565d2a1.

-----

Fix https://github.com/brave/brave-browser/issues/6513

`Reopen closed tab` and `Bookmark all tabs` items are restored to tab context menu.

Chromium removed some items from tab context menu by below CL.

CL: https://chromium-review.googlesource.com/c/chromium/src/+/1761666
Issue: https://bugs.chromium.org/p/chromium/issues/detail?id=515930

Update tab and frame context menus to match most recent UX specs.

This removes four entries from the tab context menu; changes the text on several
others to stop saying "tab(s)" explicitly; adds one entry to the frame context
menu; and changes the bookmark-related menu entry strings from "page" to "tab"
for consistency.

Most of the file changes here are due to renaming enums/APIs to match the string
changes.

image

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
Enabled state of `Bookmark all tabs` and `Reopen closed tab` are same as frame's same context menu.

`yarn test brave_browser_tests --filter=*BraveTabMenuModelTest*`

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by simonhong** [comment]
@bridiver , I think using separate proxy object makes more confusing.
So, BraveTabMenuModel implements delegate api for proxying.
Implementing delegate by subclass of `SimpleMenuModel` is common pattern in chromium code base.

Regarding to ownership, I think it's not complex anymore.
Accessing |delegate_| which is `TabContextMenuContents` is safe because `BraveTabMenuModel` is destroyed before |delegate_| is destroyed.

The reason what I choose proxying is our menu items could be added by reusing upstream code(`TabContextMenuContents) and w/o modifying their logic. WDYT?

**Comment by simonhong** [comment]
BTW, chromium re-added `Close Other Tabs` menu in c80. (https://chromium-review.googlesource.com/c/chromium/src/+/1894758)
I think that menu item is what most users concern.
Do we still need this PR? or waiting to c80?
@rebron @bsclifton WDYT?

**Comment by bsclifton** [comment]
@simonhong can you rebase this after https://github.com/brave/brave-core/pull/4009 and https://github.com/brave/brave-browser/pull/6990 get merged?

Looks like with the bump to 78.0.3904.108, the menu change you mentioned was pulled in 😄 So this PR would only need to add back the "close to left/right"

**Comment by simonhong** [comment]
@bsclifton Yup, will add other menus after merging it 👍

**Comment by simonhong** [comment]
Re-implemented by subclassing `BrowserTabStripController::ShowContextMenuForTab()`.
and this PR only added two more items restore cloased tab and bookmark all tabs.
Close other tabs will be added by upstream change after our next c78 bump.
@bridiver  PTAL again!

**Comment by simonhong** [comment]
All CI passed. set `CI/Skip`.

File: browser/ui/BUILD.gn [hunk 1]
```diff
@@ -173,6 +177,7 @@ source_set("ui") {
     "//chrome/common",
     "//components/gcm_driver:gcm_buildflags",
     "//components/prefs",
+    "//components/sessions",
```
Comment (bridiver) [inline browser/ui/BUILD.gn]: looks like we also need `//ui/views/` or is that already here somewhere?

Comment (simonhong) [reply browser/ui/BUILD.gn]: Added.

**Comment by simonhong** [comment]
Only network-audit was failed.

**Pull request merged**